File: Daily/PY/Leetcode1319-联通网络的操作次数.py
from typing import List
class UnionFind:
    def __init__(self,n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.Area = n
    def find(self,x):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def union(self,x,y):
        fx , fy = self.find(x),self.find(y)
        if fx == fy:
            return 
        if self.size[fx] < self.size[fy]:
            fx,fy = fy,fx
        self.Area -= 1
        self.parent[fy] = self.parent[fx]
        self.size[fx] += self.size[fy]
    
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        uf = UnionFind(n)
        if len(connections) < n - 1:
            return -1
        for conn in connections:
            uf.union(conn[1],conn[0])

        return uf.Area - 1

File: Daily/sbgo/Leetcode1319-联通网络的操作次数.go
package sbgo

func makeConnected(n int, connections [][]int) int {
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	size := make([]int, n)
	for i := range size {
		size[i] = 1
	}
	Area := n
	var find func(i int) int
	find = func(i int) int {
		if parent[i] == i {
			return parent[i]
		}
		parent[i] = find(parent[i])
		return parent[i]
	}
	union := func(x, y int) {
		fx, fy := find(x), find(y)
		if fx == fy {
			return
		}
		if size[fx] < size[fy] {
			fx, fy = fy, fx
		}
		Area--
		parent[fy] = fx
		size[fx] += size[fy]
	}
	mp := make(map[int]int)
	for _, conn := range connections {
		mp[conn[0]]++
		mp[conn[1]]++
		if find(conn[0]) == find(conn[1]) {
			continue
		}
		union(conn[0], conn[1])
	}
	if Area == 1 {
		return 0
	}
	m := len(mp)
	edges := len(connections)
	if edges-Area+1 >= m-1 {
		return Area - 1
	}
	return -1
}
